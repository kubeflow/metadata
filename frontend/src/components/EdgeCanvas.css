.edgeCanvas {
  position: absolute;
  width: 160px;
  height: 440px;
  border: 0;
  display: inline-block;
  margin-left: 260px;
  margin-top: 74px;
  z-index: -1;
}
.edgeCanvas.reverse {
  margin-left: 0;
  transform: translateX(-100%);
}

.kWKfgJ {
  display: block;
  overflow: visible;
  position: absolute;
}