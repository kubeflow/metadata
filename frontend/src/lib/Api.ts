import {MetadataServiceApi} from '../apis/service';
import {
  MetadataStoreServiceClient,
  ServiceError, UnaryResponse
} from "../generated/src/apis/metadata/metadata_store_service_pb_service";

/** Known Artifact properties */
export enum ArtifactProperties {
  ALL_META = '__ALL_META__',
  CREATE_TIME = 'create_time',
  DESCRIPTION = 'description',
  NAME = 'name',
  PIPELINE_NAME = 'pipeline_name',
  VERSION = 'version',
}

/** Known Artifact custom properties */
export enum ArtifactCustomProperties {
  WORKSPACE = '__kf_workspace__',
  RUN = '__kf_run__',
}

/** Known Execution properties */
export enum ExecutionProperties {
  NAME = 'name',
  PIPELINE_NAME = 'pipeline_name',
  STATE = 'state',
}

/** Known Execution custom properties */
export enum ExecutionCustomProperties {
  WORKSPACE = '__kf_workspace__',
}

/** Format for a List request */
export interface ListRequest {
  filter?: string;
  orderAscending?: boolean;
  pageSize?: number;
  pageToken?: string;
  sortBy?: string;
}

type Callback<R> = (err: ServiceError | null, res: R | null) => void;
type MetadataApiMethod<T, R> = (request: T, callback: Callback<R>) => UnaryResponse;
type PromiseBasedMetadataApiMethod<T, R> = (request: T) => Promise<{ response: R | null, error: ServiceError | null }>;

/**
 * Converts a callback based api method to promise based api method.
 */
function makePromiseApi<T, R>(apiMethod: MetadataApiMethod<T, R>): PromiseBasedMetadataApiMethod<T, R> {
  return (request: T) => new Promise((resolve, reject) => {
    const handler = (error: ServiceError | null, response: R | null) => {
      // resolve both response and error to keep type information
      resolve({ response, error });
    };
    apiMethod(request, handler);
  });
}

const metadataServiceClient = new MetadataStoreServiceClient('');

// TODO: add all other api methods we need here.
const metadataServicePromiseClient = {
  getArtifactTypes: makePromiseApi(metadataServiceClient.getArtifactTypes.bind(metadataServiceClient)),
  getArtifactsByID: makePromiseApi(metadataServiceClient.getArtifactsByID.bind(metadataServiceClient)),
  getEventsByArtifactIDs: makePromiseApi(metadataServiceClient.getEventsByArtifactIDs.bind(metadataServiceClient)),
  getEventsByExecutionIDs: makePromiseApi(metadataServiceClient.getEventsByExecutionIDs.bind(metadataServiceClient)),
  getExecutionsByID: makePromiseApi(metadataServiceClient.getExecutionsByID.bind(metadataServiceClient)),
};

/**
 * Class to wrap backend APIs.
 */
export class Api {

  private static instance: Api;
  private metadataServiceApi: MetadataServiceApi;

  /**
   * Factory function to return an Api instance.
   */
  public static getInstance(): Api {
    if (!Api.instance) {
      const location = window.location;
      let path = `${location.protocol}//${location.host}${location.pathname}`;
      if (path.endsWith('/')) path = path.slice(0, path.length - 1);
      Api.instance = new Api(path);
    }
    return Api.instance;
  }

  private constructor(basePath: string) {
    this.metadataServiceApi = new MetadataServiceApi({basePath});
  }

  get metadataService(): MetadataServiceApi {
    return this.metadataServiceApi;
  }

  get metadataStoreService() {
    return metadataServicePromiseClient;
  }
}
