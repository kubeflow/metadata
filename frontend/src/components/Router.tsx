/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import Banner, {BannerProps} from '../components/Banner';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogTitle from '@material-ui/core/DialogTitle';
import Page404 from '../pages/404';
import Snackbar, {SnackbarProps} from '@material-ui/core/Snackbar';
import Toolbar, {ToolbarProps} from './Toolbar';
import {Route, Switch, Redirect, HashRouter} from 'react-router-dom';
import {classes, stylesheet} from 'typestyle';
import {commonCss} from '../Css';
import ArtifactList from '../pages/ArtifactList';
import ArtifactDetails from '../pages/ArtifactDetails';
import ExecutionList from '../pages/ExecutionList';
import SideNav from '../pages/SideNav';

const css = stylesheet({
  dialog: {
    minWidth: 250,
  },
});

export enum RouteParams {
  ARTIFACT_TYPE = 'artifactType',
  EXECUTION_TYPE = 'executionType',
  ID = 'id',
}

export const RoutePage = {
  ARTIFACTS: '/artifacts',
  ARTIFACT_DETAILS: `/artifact_types/:${RouteParams.ARTIFACT_TYPE}+/artifacts/:${RouteParams.ID}`,
  EXECUTIONS: '/executions',
  EXECUTION_DETAILS: `/execution_types/:${RouteParams.EXECUTION_TYPE}+/executions/:${RouteParams.ID}`,
};

export interface DialogProps {
  buttons?: Array<{onClick?: () => any, text: string}>;
  // TODO: This should be generalized to any react component.
  content?: string;
  onClose?: () => any;
  open?: boolean;
  title?: string;
}

interface RouteConfig {
  path: string,
  Component: React.ComponentClass,
  view?: any
}

interface RouteComponentState {
  bannerProps: BannerProps;
  dialogProps: DialogProps;
  snackbarProps: SnackbarProps;
  toolbarProps: ToolbarProps;
}

class Router extends React.Component<{}, RouteComponentState> {

  private routes: RouteConfig[] = [
    {path: RoutePage.ARTIFACTS, Component: ArtifactList},
    {path: RoutePage.ARTIFACT_DETAILS, Component: ArtifactDetails},
    {path: RoutePage.EXECUTIONS, Component: ExecutionList},
    {path: RoutePage.EXECUTION_DETAILS, Component: ArtifactDetails},
  ];

  constructor(props: any) {
    super(props);
    this.state = {
      bannerProps: {},
      dialogProps: {open: false},
      snackbarProps: {autoHideDuration: 5000, open: false},
      toolbarProps: {breadcrumbs: [{displayName: '', href: ''}], actions: [], ...props},
    };
  }

  public render(): JSX.Element {
    const childProps = {
      toolbarProps: this.state.toolbarProps,
      updateBanner: this._updateBanner.bind(this),
      updateDialog: this._updateDialog.bind(this),
      updateSnackbar: this._updateSnackbar.bind(this),
      updateToolbar: this._updateToolbar.bind(this),
    };
    return (
      <HashRouter>
        <div className={commonCss.page}>
          <div className={commonCss.flexGrow}>
          <Route render={({ ...props }) => (<SideNav page={props.location.pathname} {...props} />)} />
            <div className={classes(commonCss.page)}>
              <Route render={({...props}) => (<Toolbar {...this.state.toolbarProps} {...props} />)} />
              {this.state.bannerProps.message
                && <Banner
                  message={this.state.bannerProps.message}
                  mode={this.state.bannerProps.mode}
                  additionalInfo={this.state.bannerProps.additionalInfo}
                  refresh={this.state.bannerProps.refresh} />}
              <Switch>
                <Route exact={true} path={'/'} render={({...props}) => (
                  <Redirect to={RoutePage.ARTIFACTS} {...props} />
                )} />
                {this.routes.map((route, i) => {
                  const {path, Component, ...otherProps} = {...route};
                  return <Route key={i} exact={true} path={path} render={({...props}) => (
                    <Component {...props} {...childProps} {...otherProps} />
                  )} />;
                })}

                {/* 404 */}
                {<Route render={({...props}) => <Page404 {...props} {...childProps} />} />}
              </Switch>

              <Snackbar
                autoHideDuration={this.state.snackbarProps.autoHideDuration}
                message={this.state.snackbarProps.message}
                open={this.state.snackbarProps.open}
                onClose={this._handleSnackbarClose.bind(this)}
              />
            </div>
          </div>

          <Dialog open={this.state.dialogProps.open !== false} classes={{paper: css.dialog}}
            className='dialog' onClose={() => this._handleDialogClosed()}>
            {this.state.dialogProps.title && (
              <DialogTitle> {this.state.dialogProps.title}</DialogTitle>
            )}
            {this.state.dialogProps.content && (
              <DialogContent className={commonCss.prewrap}>
                {this.state.dialogProps.content}
              </DialogContent>
            )}
            {this.state.dialogProps.buttons && (
              <DialogActions>
                {this.state.dialogProps.buttons.map((b, i) =>
                  <Button key={i} onClick={() => this._handleDialogClosed(b.onClick)}
                    className='dialogButton' color='secondary'>
                    {b.text}
                  </Button>)}
              </DialogActions>
            )}
          </Dialog>
        </div>
      </HashRouter>
    );
  }

  private _updateDialog(dialogProps: DialogProps): void {
    // Assuming components will want to open the dialog by defaut.
    if (dialogProps.open === undefined) {
      dialogProps.open = true;
    }
    this.setState({dialogProps});
  }

  private _handleDialogClosed(onClick?: () => void): void {
    this.setState({dialogProps: {open: false}});
    if (onClick) {
      onClick();
    }
    if (this.state.dialogProps.onClose) {
      this.state.dialogProps.onClose();
    }
  }

  private _updateToolbar(newToolbarProps: Partial<ToolbarProps>): void {
    const toolbarProps = Object.assign(this.state.toolbarProps, newToolbarProps);
    this.setState({toolbarProps});
  }

  private _updateBanner(bannerProps: BannerProps): void {
    this.setState({bannerProps});
  }

  private _updateSnackbar(snackbarProps: SnackbarProps): void {
    snackbarProps.autoHideDuration =
      snackbarProps.autoHideDuration || this.state.snackbarProps.autoHideDuration;
    this.setState({snackbarProps});
  }

  private _handleSnackbarClose(): void {
    this.setState({snackbarProps: {open: false, message: ''}});
  }
}

// TODO: loading/error experience until backend is reachable

export default Router;
