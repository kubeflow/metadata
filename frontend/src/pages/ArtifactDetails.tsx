/*
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the 'License');
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an 'AS IS' BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import {Page} from './Page';
import {ToolbarProps} from '../components/Toolbar';
import {RoutePage, RouteParams} from '../components/Router';
import {Api, ArtifactProperties} from '../lib/Api';
import {MlMetadataArtifact} from '../apis/service';
import {classes} from 'typestyle';
import {commonCss, padding} from '../Css';
import {CircularProgress} from '@material-ui/core';
import {titleCase, getResourceProperty} from '../lib/Utils';
import {ResourceInfo} from '../components/ResourceInfo';
import MD2Tabs from '../atoms/MD2Tabs';

export enum ArtifactDetailsTab {
  OVERVIEW = 0,
  LINEAGE_EXPLORER = 1,
  DEPLOYMENTS = 2
}

const tabs = {
  [ArtifactDetailsTab.OVERVIEW]: {name: 'Overview'},
  [ArtifactDetailsTab.LINEAGE_EXPLORER]: {name: 'Lineage Explorer'},
  [ArtifactDetailsTab.DEPLOYMENTS]: {name: 'Deployments'}
};

const tabNames = Object.values(tabs).map(tabConfig => tabConfig.name);

interface ArtifactDetailsState {
  artifact?: MlMetadataArtifact;
  selectedTab: ArtifactDetailsTab;
}

export default class ArtifactDetails extends Page<{}, ArtifactDetailsState> {
  private api = Api.getInstance();

  constructor(props: {}) {
    super(props);
    this.state = {
      selectedTab: ArtifactDetailsTab.OVERVIEW
    };
    this.load = this.load.bind(this);
  }

  private get fullTypeName(): string {
    return this.props.match.params[RouteParams.ARTIFACT_TYPE] || '';
  }

  private get properTypeName(): string {
    const parts = this.fullTypeName.split('/');
    if (!parts.length) return '';

    return titleCase(parts[parts.length - 1]);
  }

  private get id(): string {
    return this.props.match.params[RouteParams.ID];
  }

  public async componentDidMount(): Promise<void> {
    return this.load();
  }

  public render(): JSX.Element {
    if (!this.state.artifact) return <CircularProgress />;
    return (
      <div className={classes(commonCss.page)}>
        <div className={classes(padding(20, 'tb'))}>
          <MD2Tabs
            tabs={tabNames}
            selectedTab={this.state.selectedTab}
            onSwitch={this.switchTab.bind(this)}
          />
        </div>
        <div className={classes(padding(20, 'lr'))}>
          {this.state.selectedTab === ArtifactDetailsTab.OVERVIEW && (
            <ResourceInfo typeName={this.properTypeName} resource={this.state.artifact} />
          )}
          {this.state.selectedTab === ArtifactDetailsTab.LINEAGE_EXPLORER && (
            <span>Lineage Explorer</span>
          )}
          {this.state.selectedTab === ArtifactDetailsTab.DEPLOYMENTS && <span>Deployments</span>}
        </div>
      </div>
    );
  }

  public getInitialToolbarState(): ToolbarProps {
    return {
      actions: [],
      breadcrumbs: [{displayName: 'Artifacts', href: RoutePage.ARTIFACTS}],
      pageTitle: `${this.properTypeName} ${this.id} details`
    };
  }

  public async refresh(): Promise<void> {
    return this.load();
  }

  private async load(): Promise<void> {
    try {
      const {artifact} = await this.api.metadataService.getArtifact(this.id, this.fullTypeName);
      if (!artifact) {
        throw new Error(`No ${this.fullTypeName} identified by id: ${this.id}`);
      }

      const artifactName = getResourceProperty(artifact, ArtifactProperties.NAME);
      let title = artifactName ? artifactName.toString() : '';
      const version = getResourceProperty(artifact, ArtifactProperties.VERSION);
      if (version) {
        title += ` (version: ${version})`;
      }
      this.props.updateToolbar({
        pageTitle: title
      });
      this.setState({artifact});
    } catch (err) {
      this.showPageError(`Unable to retrieve ${this.fullTypeName} ${this.id}.`, err);
    }
  }

  private switchTab(selectedTab: number) {
    this.setState({
      selectedTab
    });
  }
}
