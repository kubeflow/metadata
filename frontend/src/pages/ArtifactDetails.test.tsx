import * as React from 'react';
import ArtifactDetails, {ArtifactDetailsTab} from './ArtifactDetails';
import {PageProps} from './Page';
import {shallow, ShallowWrapper, ReactWrapper} from 'enzyme';
import {Api, ArtifactCustomProperties, ArtifactProperties} from '../lib/Api';
import * as TestUtils from '../TestUtils'
import {RouteParams} from '../components/Router';
import {serviceError, stringValue} from '../TestUtils';
import {GetArtifactsByIDResponse} from '../generated/src/apis/metadata/metadata_store_service_pb';
import {Artifact} from '../generated/src/apis/metadata/metadata_store_pb';

describe('ArtifactDetails', () => {
  let tree: ShallowWrapper | ReactWrapper;
  const updateBannerSpy = jest.fn();
  const updateToolbarSpy = jest.fn();
  const historyPushSpy = jest.fn();
  const mockGetArtifact = jest.spyOn(Api.getInstance().metadataStoreService, 'getArtifactsByID');
  const fakeGetArtifactByIDResponse = new GetArtifactsByIDResponse();
  const artifact = new Artifact();
  artifact.setId(1);
  artifact.setTypeId(1);
  artifact.setUri('gs://my-bucket/mnist');
  artifact.getPropertiesMap().set(ArtifactProperties.NAME, stringValue('test model'));
  artifact.getPropertiesMap().set(ArtifactProperties.DESCRIPTION, stringValue('A really great model'));
  artifact.getPropertiesMap().set(ArtifactProperties.VERSION, stringValue('v1'));
  artifact.getPropertiesMap().set(ArtifactProperties.CREATE_TIME, stringValue('2019-06-12T01:21:48.259263Z'));
  artifact.getPropertiesMap().set(ArtifactProperties.ALL_META, stringValue(
      '{"hyperparameters": {"early_stop": true, ' +
      '"layers": [10, 3, 1], "learning_rate": 0.5}, ' +
      '"model_type": "neural network", ' +
      '"training_framework": {"name": "tensorflow", "version": "v1.0"))'));
  artifact.getCustomPropertiesMap().set(ArtifactCustomProperties.WORKSPACE, stringValue('workspace-1'));
  artifact.getCustomPropertiesMap().set(ArtifactCustomProperties.RUN, stringValue('1'));
  fakeGetArtifactByIDResponse.addArtifacts(artifact);

  const MODEL_TYPE = 'kubeflow.org/alpha/model';
  const FAKE_MODEL_ID = '1';

  function generateProps(): PageProps {
    return TestUtils.generatePageProps(
      ArtifactDetails,
      {} as any,
      {
        params: {
          [RouteParams.ARTIFACT_TYPE]: MODEL_TYPE,
          [RouteParams.ID]: FAKE_MODEL_ID
        }
      } as any,
      historyPushSpy,
      updateBannerSpy,
      jest.fn(),
      updateToolbarSpy,
      jest.fn());
  }

  beforeEach(() => {
    jest.clearAllMocks();
  });

  afterEach(() => tree.unmount());

  it('Renders progress bar with no data', () => {
    tree = shallow(<ArtifactDetails {...generateProps()} />);

    expect(tree).toMatchSnapshot();
  });

  it('Renders with a Model Artifact and updates the page title', async () => {
    mockGetArtifact.mockResolvedValue({
      error: null,
      response: fakeGetArtifactByIDResponse,
    });
    tree = TestUtils.mountWithRouter(<ArtifactDetails {...generateProps()} />);

    await mockGetArtifact;
    await TestUtils.flushPromises();
    tree.update();
    expect(tree).toMatchSnapshot();
    expect(updateToolbarSpy).toHaveBeenLastCalledWith(expect.objectContaining({
      pageTitle: 'test model (version: v1)',
    }));
  });

  it('Shows error when returned Artifact is empty', async () => {
    mockGetArtifact.mockResolvedValue({
      error: null,
      response: new GetArtifactsByIDResponse(),
    });
    tree = TestUtils.mountWithRouter(<ArtifactDetails {...generateProps()} />);

    await mockGetArtifact;
    await TestUtils.flushPromises();
    expect(updateBannerSpy).toHaveBeenCalledWith(expect.objectContaining({
      message: 'No kubeflow.org/alpha/model identified by id: 1',
      mode: 'error',
    }));
  });

  it('Shows error when Artifact cannot be retrieved', async () => {
    mockGetArtifact.mockResolvedValue({
      error: serviceError,
      response: fakeGetArtifactByIDResponse,
    });
    tree = TestUtils.mountWithRouter(<ArtifactDetails {...generateProps()} />);

    await mockGetArtifact;
    await TestUtils.flushPromises();
    expect(updateBannerSpy).toHaveBeenCalledWith(expect.objectContaining({
      message: 'Unable to retrieve kubeflow.org/alpha/model 1. ' +
        'Click Details for more information.',
      mode: 'error',
    }));
  });

  it('Renders the Overview tab for an artifact', async () => {
    mockGetArtifact.mockResolvedValue({
      error: null,
      response: fakeGetArtifactByIDResponse,
    });
    tree = TestUtils.mountWithRouter(<ArtifactDetails {...generateProps()} />);
    tree.setState({
      selectedTab: ArtifactDetailsTab.OVERVIEW
    });

    await mockGetArtifact;
    await TestUtils.flushPromises();
    tree.update();
    expect(tree).toMatchSnapshot();
  });

  it('Renders the Lineage Explorer tab for an artifact', async () => {
    mockGetArtifact.mockResolvedValue({
      error: null,
      response: fakeGetArtifactByIDResponse,
    });
    tree = TestUtils.mountWithRouter(<ArtifactDetails {...generateProps()} />);
    tree.setState({
      selectedTab: ArtifactDetailsTab.LINEAGE_EXPLORER
    });

    await mockGetArtifact;
    await TestUtils.flushPromises();
    tree.update();
    expect(tree).toMatchSnapshot();
  });

  it('Renders the Deployments tab for an artifact', async () => {
    mockGetArtifact.mockResolvedValue({
      error: null,
      response: fakeGetArtifactByIDResponse,
    });
    tree = TestUtils.mountWithRouter(<ArtifactDetails {...generateProps()} />);
    tree.setState({
      selectedTab: ArtifactDetailsTab.DEPLOYMENTS
    });

    await mockGetArtifact;
    await TestUtils.flushPromises();
    tree.update();
    expect(tree).toMatchSnapshot();
  });
});
