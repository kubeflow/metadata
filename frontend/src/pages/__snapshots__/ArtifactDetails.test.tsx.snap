// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArtifactDetails Renders progress bar with no data 1`] = `<WithStyles(CircularProgress) />`;

exports[`ArtifactDetails Renders with a Model Artifact and updates the page title 1`] = `
<ArtifactDetails
  history={
    Object {
      "push": [MockFunction],
    }
  }
  location={Object {}}
  match={
    Object {
      "params": Object {
        "artifactType": "kubeflow.org/alpha/model",
        "id": "1",
      },
    }
  }
  toolbarProps={
    Object {
      "actions": Array [],
      "breadcrumbs": Array [
        Object {
          "displayName": "Artifacts",
          "href": "/artifacts",
        },
      ],
      "pageTitle": "Model 1 details",
    }
  }
  updateBanner={[MockFunction]}
  updateDialog={[MockFunction]}
  updateSnackbar={[MockFunction]}
  updateToolbar={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "actions": Array [],
            "breadcrumbs": Array [
              Object {
                "displayName": "Artifacts",
                "href": "/artifacts",
              },
            ],
            "pageTitle": "Model 1 details",
          },
        ],
        Array [
          Object {
            "pageTitle": "test model (version: v1)",
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  <div
    className="page"
  >
    <ResourceInfo
      resource={
        Object {
          "array": Array [
            1,
            1,
            "gs://my-bucket/mnist",
            Array [
              Array [
                "__ALL_META__",
                Array [
                  undefined,
                  undefined,
                  "{\\"hyperparameters\\": {\\"early_stop\\": true, \\"layers\\": [10, 3, 1], \\"learning_rate\\": 0.5}, \\"model_type\\": \\"neural network\\", \\"training_framework\\": {\\"name\\": \\"tensorflow\\", \\"version\\": \\"v1.0\\"))",
                ],
              ],
              Array [
                "create_time",
                Array [
                  undefined,
                  undefined,
                  "2019-06-12T01:21:48.259263Z",
                ],
              ],
              Array [
                "description",
                Array [
                  undefined,
                  undefined,
                  "A really great model",
                ],
              ],
              Array [
                "name",
                Array [
                  undefined,
                  undefined,
                  "test model",
                ],
              ],
              Array [
                "version",
                Array [
                  undefined,
                  undefined,
                  "v1",
                ],
              ],
            ],
            Array [
              Array [
                "__kf_run__",
                Array [
                  undefined,
                  undefined,
                  "1",
                ],
              ],
              Array [
                "__kf_workspace__",
                Array [
                  undefined,
                  undefined,
                  "workspace-1",
                ],
              ],
            ],
          ],
          "arrayIndexOffset_": -1,
          "convertedPrimitiveFields_": Object {},
          "messageId_": undefined,
          "pivot_": 1.7976931348623157e+308,
          "wrappers_": Object {
            "4": Object {
              "arrClean": true,
              "arr_": Array [
                Array [
                  "__ALL_META__",
                  Array [
                    undefined,
                    undefined,
                    "{\\"hyperparameters\\": {\\"early_stop\\": true, \\"layers\\": [10, 3, 1], \\"learning_rate\\": 0.5}, \\"model_type\\": \\"neural network\\", \\"training_framework\\": {\\"name\\": \\"tensorflow\\", \\"version\\": \\"v1.0\\"))",
                  ],
                ],
                Array [
                  "create_time",
                  Array [
                    undefined,
                    undefined,
                    "2019-06-12T01:21:48.259263Z",
                  ],
                ],
                Array [
                  "description",
                  Array [
                    undefined,
                    undefined,
                    "A really great model",
                  ],
                ],
                Array [
                  "name",
                  Array [
                    undefined,
                    undefined,
                    "test model",
                  ],
                ],
                Array [
                  "version",
                  Array [
                    undefined,
                    undefined,
                    "v1",
                  ],
                ],
              ],
              "map_": Object {
                "__ALL_META__": Object {
                  "key": "__ALL_META__",
                  "value": Array [
                    undefined,
                    undefined,
                    "{\\"hyperparameters\\": {\\"early_stop\\": true, \\"layers\\": [10, 3, 1], \\"learning_rate\\": 0.5}, \\"model_type\\": \\"neural network\\", \\"training_framework\\": {\\"name\\": \\"tensorflow\\", \\"version\\": \\"v1.0\\"))",
                  ],
                  "valueWrapper": Object {
                    "array": Array [
                      undefined,
                      undefined,
                      "{\\"hyperparameters\\": {\\"early_stop\\": true, \\"layers\\": [10, 3, 1], \\"learning_rate\\": 0.5}, \\"model_type\\": \\"neural network\\", \\"training_framework\\": {\\"name\\": \\"tensorflow\\", \\"version\\": \\"v1.0\\"))",
                    ],
                    "arrayIndexOffset_": -1,
                    "convertedPrimitiveFields_": Object {},
                    "messageId_": undefined,
                    "pivot_": 1.7976931348623157e+308,
                    "wrappers_": null,
                  },
                },
                "create_time": Object {
                  "key": "create_time",
                  "value": Array [
                    undefined,
                    undefined,
                    "2019-06-12T01:21:48.259263Z",
                  ],
                  "valueWrapper": Object {
                    "array": Array [
                      undefined,
                      undefined,
                      "2019-06-12T01:21:48.259263Z",
                    ],
                    "arrayIndexOffset_": -1,
                    "convertedPrimitiveFields_": Object {},
                    "messageId_": undefined,
                    "pivot_": 1.7976931348623157e+308,
                    "wrappers_": null,
                  },
                },
                "description": Object {
                  "key": "description",
                  "value": Array [
                    undefined,
                    undefined,
                    "A really great model",
                  ],
                  "valueWrapper": Object {
                    "array": Array [
                      undefined,
                      undefined,
                      "A really great model",
                    ],
                    "arrayIndexOffset_": -1,
                    "convertedPrimitiveFields_": Object {},
                    "messageId_": undefined,
                    "pivot_": 1.7976931348623157e+308,
                    "wrappers_": null,
                  },
                },
                "name": Object {
                  "key": "name",
                  "value": Array [
                    undefined,
                    undefined,
                    "test model",
                  ],
                  "valueWrapper": Object {
                    "array": Array [
                      undefined,
                      undefined,
                      "test model",
                    ],
                    "arrayIndexOffset_": -1,
                    "convertedPrimitiveFields_": Object {},
                    "messageId_": undefined,
                    "pivot_": 1.7976931348623157e+308,
                    "wrappers_": null,
                  },
                },
                "version": Object {
                  "key": "version",
                  "value": Array [
                    undefined,
                    undefined,
                    "v1",
                  ],
                  "valueWrapper": Object {
                    "array": Array [
                      undefined,
                      undefined,
                      "v1",
                    ],
                    "arrayIndexOffset_": -1,
                    "convertedPrimitiveFields_": Object {},
                    "messageId_": undefined,
                    "pivot_": 1.7976931348623157e+308,
                    "wrappers_": null,
                  },
                },
              },
              "valueCtor_": [Function],
            },
            "5": Object {
              "arrClean": true,
              "arr_": Array [
                Array [
                  "__kf_run__",
                  Array [
                    undefined,
                    undefined,
                    "1",
                  ],
                ],
                Array [
                  "__kf_workspace__",
                  Array [
                    undefined,
                    undefined,
                    "workspace-1",
                  ],
                ],
              ],
              "map_": Object {
                "__kf_run__": Object {
                  "key": "__kf_run__",
                  "value": Array [
                    undefined,
                    undefined,
                    "1",
                  ],
                  "valueWrapper": Object {
                    "array": Array [
                      undefined,
                      undefined,
                      "1",
                    ],
                    "arrayIndexOffset_": -1,
                    "convertedPrimitiveFields_": Object {},
                    "messageId_": undefined,
                    "pivot_": 1.7976931348623157e+308,
                    "wrappers_": null,
                  },
                },
                "__kf_workspace__": Object {
                  "key": "__kf_workspace__",
                  "value": Array [
                    undefined,
                    undefined,
                    "workspace-1",
                  ],
                  "valueWrapper": Object {
                    "array": Array [
                      undefined,
                      undefined,
                      "workspace-1",
                    ],
                    "arrayIndexOffset_": -1,
                    "convertedPrimitiveFields_": Object {},
                    "messageId_": undefined,
                    "pivot_": 1.7976931348623157e+308,
                    "wrappers_": null,
                  },
                },
              },
              "valueCtor_": [Function],
            },
          },
        }
      }
      typeName="Model"
    >
      <section>
        <h1
          className="header"
        >
          Type: 
          Model
        </h1>
        <h2
          className="header2"
        >
          Properties
        </h2>
        <dl
          className="resourceInfo"
        >
          <div
            className="field"
            key="create_time"
          >
            <dt
              className="term"
            >
              create_time
            </dt>
            <dd
              className="value"
            >
              2019-06-12T01:21:48.259263Z
            </dd>
          </div>
          <div
            className="field"
            key="description"
          >
            <dt
              className="term"
            >
              description
            </dt>
            <dd
              className="value"
            >
              A really great model
            </dd>
          </div>
          <div
            className="field"
            key="name"
          >
            <dt
              className="term"
            >
              name
            </dt>
            <dd
              className="value"
            >
              test model
            </dd>
          </div>
          <div
            className="field"
            key="version"
          >
            <dt
              className="term"
            >
              version
            </dt>
            <dd
              className="value"
            >
              v1
            </dd>
          </div>
        </dl>
        <h2
          className="header2"
        >
          Custom Properties
        </h2>
        <dl
          className="resourceInfo"
        >
          <div
            className="field"
            key="__kf_run__"
          >
            <dt
              className="term"
            >
              __kf_run__
            </dt>
            <dd
              className="value"
            >
              1
            </dd>
          </div>
          <div
            className="field"
            key="__kf_workspace__"
          >
            <dt
              className="term"
            >
              __kf_workspace__
            </dt>
            <dd
              className="value"
            >
              workspace-1
            </dd>
          </div>
        </dl>
      </section>
    </ResourceInfo>
  </div>
</ArtifactDetails>
`;
